### ISSUE TITLE
Custom Compression Provider is not being triggered

### ISSUE THREAD

**Issue by alexvazquez** [grpc/grpc-dotnet#1106]
I created a new Core API Project and I had the Interface and class for custom compression as it states here:

 https://github.com/grpc/grpc-dotnet/blob/c48ae4442fdae1881622ad94d02dfda45912493a/src/Grpc.Net.Common/Compression/GzipCompressionProvider.cs

Then in my Startup.cs class I have the following:

```
services.AddGrpc().AddServiceOptions<GreeterService>(options =>
            {
                options.ResponseCompressionLevel = System.IO.Compression.CompressionLevel.Optimal;
                options.ResponseCompressionAlgorithm = "gzip";
                options.CompressionProviders = new List<ICompressionProvider>();
            });
```
Then I set a few breakpoints just to see if the new provider is being hit without any success:

![image](https://user-images.githubusercontent.com/347747/98488410-9c1cd380-21ee-11eb-8328-995f3d57c3ef.png)

Any clue on how to make this work? Thanks

**Comment by JamesNK** [comment]
Why are you not registering the new compression provider with `GrpcChannelOptions.CompressionProviders`?

**Comment by alexvazquez** [comment]
> Why are you not registering the new compression provider with `GrpcChannelOptions.CompressionProviders`?

How can I use `GrpcChannelOptions.CompressionProvider` at the server level? Didn't find any code to do that.

On the client-side, because I'm using .NET Framework 4.6 I can't use the .NET Core dependency so my dependencies are:

```
<packages>
  <package id="Google.Protobuf" version="3.13.0" targetFramework="net461" />
  <package id="Grpc" version="2.33.1" targetFramework="net461" />
  <package id="Grpc.Core" version="2.33.1" targetFramework="net461" />
  <package id="Grpc.Core.Api" version="2.33.1" targetFramework="net461" />
  <package id="Grpc.Tools" version="2.33.1" targetFramework="net461" developmentDependency="true" />
  <package id="System.Buffers" version="4.4.0" targetFramework="net461" />
  <package id="System.Memory" version="4.5.3" targetFramework="net461" />
  <package id="System.Numerics.Vectors" version="4.4.0" targetFramework="net461" />
  <package id="System.Runtime.CompilerServices.Unsafe" version="4.5.2" targetFramework="net461" />
</packages>

```
 And my code looks like this:

```

 string s = GetRootCertificates();
            var channel_creds = new SslCredentials(s);

            Channel channel = new Channel("localhost:5001", channel_creds, new ChannelOption[] {
                new ChannelOption(ChannelOptions.MaxReceiveMessageLength, 10 * 1024 * 1024), //plenty of room
                new ChannelOption("grpc.default_compression_algorithm",2),
                new ChannelOption("grpc.default_compression_level",(int)CompressionLevel.High),
            });

            var invoker = channel.Intercept(new ClientLoggerInterceptor());

            var client = new Greeter.GreeterClient(invoker);

            var headers = new Metadata();
            headers.Add("grpc-internal-encoding-request", "gzip");
            headers.Add("requester", "client");

            var callOptions = new CallOptions().WithHeaders(headers).WithDeadline(DateTime.UtcNow.AddSeconds(15)).WithWaitForReady(false);

            DataSetResponse response = await client.GetPatientInfoAsync(new PatientInfoRequest { Uid = "DF24387B-815D-4E06-AD76-9E9E3C255118", LoadContractsOnDemand = false }, callOptions);
            
```
I haven't find any sample on how to do custom providers for `Channel channel = new Channel(`

**Comment by JamesNK** [comment]
Docs: https://docs.microsoft.com/en-us/aspnet/core/grpc/configuration

Channel is Grpc.Core. This repo is for Grpc.Net.Client.

**Comment by alexvazquez** [comment]
> Docs: https://docs.microsoft.com/en-us/aspnet/core/grpc/configuration
> 
> Channel is Grpc.Core. This repo is for Grpc.Net.Client.

@JamesNK but Im using  `Grpc.Net.Client `on server side and cant find an example on how to use the CustomProviders. Any link where I can see something please?

**Comment by alexvazquez** [comment]
I just solved it! Good now I can get my custom compression tool to work. Now I will have to implement something similar client side using Grpc.Core.

**Comment by chwarr** [comment]
@alexvazquez, can you summarize how you solved it here so that anyone who reads this in the future can learn as well?

**Comment by alexvazquez** [comment]
The way I solve it was to add the following code:

```
services.AddGrpc().AddServiceOptions<GreeterService>(options =>
            {

                options.CompressionProviders = new List<ICompressionProvider>()
                {
                    new GrpcService1.Compression.CustomGzipCompressionProvider(System.IO.Compression.CompressionLevel.Optimal)
                };
            
                options.EnableDetailedErrors = true;
            });
```

And my custom Gzip compression for now is exactly as the GzipCompressionProvider.
```

public class CustomGzipCompressionProvider : ICompressionProvider
    {
        private readonly CompressionLevel _defaultCompressionLevel;

        /// <summary>
        /// Initializes a new instance of the <see cref="CustomGzipCompressionProvider"/> class with the specified <see cref="CompressionLevel"/>.
        /// </summary>
        /// <param name="defaultCompressionLevel">The default compression level to use when compressing data.</param>
        public CustomGzipCompressionProvider(CompressionLevel defaultCompressionLevel)
        {
            _defaultCompressionLevel = defaultCompressionLevel;
        }

        /// <summary>
        /// The encoding name used in the 'grpc-encoding' and 'grpc-accept-encoding' request and response headers.
        /// </summary>
        public string EncodingName => "gzip";

        /// <summary>
        /// Create a new compression stream.
        /// </summary>
        /// <param name="stream">The stream that compressed data is written to.</param>
        /// <param name="compressionLevel">The compression level.</param>
        /// <returns>A stream used to compress data.</returns>
        public Stream CreateCompressionStream(Stream stream, CompressionLevel? compressionLevel)
        {
            return new GZipStream(stream, compressionLevel ?? _defaultCompressionLevel, leaveOpen: true);
        }

        /// <summary>
        /// Create a new decompression stream.
        /// </summary>
        /// <param name="stream">The stream that compressed data is copied from.</param>
        /// <returns>A stream used to decompress data.</returns>
        public Stream CreateDecompressionStream(Stream stream)
        {
            return new GZipStream(stream, CompressionMode.Decompress, leaveOpen: true);
        }
    }
```

**Comment by alexvazquez** [comment]
> I just solved it! Good now I can get my custom compression tool to work. Now I will have to implement something similar client side using Grpc.Core.

How did you solve it? Thanks
